api/src/server/auth.rs

```diff
@@ -0,0 +1,140 @@
+use crate::{
+    error::ResponseError,
+    models::{NewUser, User},
+    schema::users,
+    util::Pool,
+};
+use actix_identity::Identity;
+use actix_web::{get, http, web, HttpResponse, Responder};
+use diesel::{prelude::RunQueryDsl, PgConnection};
+use openid::{Client, DiscoveredClient, Options, Token, Userinfo};
+use reqwest::Url;
+use serde::Deserialize;
+use std::{collections::HashMap, sync::RwLock};
+pub struct Sessions {
+    pub map: HashMap<String, (User, Token, Userinfo)>,
+}
+
+pub async fn make_client(
+    client_id: String,
+    client_secret: String,
+    issuer_url: &str,
+) -> Client {
+    let redirect = Some(host("/login/oauth2/code/oidc"));
+
+    let issuer = match Url::parse(issuer_url) {
+        Ok(res) => res,
+        Err(e) => panic!(e),
+    };
+
+    match DiscoveredClient::discover(client_id, client_secret, redirect, issuer)
+        .await
+    {
+        Ok(res) => res,
+        Err(e) => panic!(e),
+    }
+}
+
+// The frontend will redirect to this before being sent off to the
+// actual openid provdider
+// TODO: add query param to say which provider to use
```

Comment:
Not a huge deal, but I think it'd be better to use a route param, like `/blah/<provider>`